proxy/envoy/config.go

```diff
@@ -163,6 +180,43 @@ func build(instances []*model.ServiceInstance, services []*model.Service,
 				},
 			}},
 		}
+
+		// Add TCP filters that share a port with HTTP filters.
+		if _, ok := tcpRouteConfigs[port]; ok {
+
+			// TCP and HTTP filter on the same port seems to cause
+			// envoy to assign all traffic to TCP filter regardless of
+			// the routing config. Temporarily drop TCP filters that
+			// share a port with HTTP (e.g. istio-egress) until this is
+			// root caused.
+
+			// listener.Filters = append(listener.Filters, &NetworkFilter{
+			// 	Type: "read",
+			// 	Name: TCPProxyFilter,
+			// 	Config: TCPProxyFilterConfig{
+			// 		StatPrefix:  "tcp",
+			// 		RouteConfig: tcpConfig,
+			// 	},
+			// })
+			// clusters = append(clusters, tcpConfig.clusters()...)
+			delete(tcpRouteConfigs, port)
+		}
+		listeners = append(listeners, listener)
+	}
+
+	for port, tcpConfig := range tcpRouteConfigs {
+		clusters = append(clusters, tcpConfig.clusters()...)
+		listener := &Listener{
+			Port: port,
```

Comment:
The commented out lines at 192-200 attempted to do a basic of merge for any TCP filter that shared a port with an HTTP by appending the TCP proxy filter after the  HTTP manager filter in the network filter array. The expectation was that envoy would stop iterating through the network filters when it found a matching HTTP filter and skip over the TCP filter but that doesn't seem to be happening despite the fact that the destination IP address shouldn't match anyway. It's quite possible I messed up the TCP route config, e.g. @rshriram's comment about using cluster IP or pod IP range.